File: app/src/main/java/email/schaal/ocreader/util/LoginError.java
package email.schaal.ocreader.util;

import android.content.Context;

import java.net.UnknownHostException;
import java.security.cert.CertificateException;

import javax.net.ssl.SSLHandshakeException;

import email.schaal.ocreader.R;

/**
 * Turn login errors into human-readable strings
 */
public class LoginError {
    public enum Section {
        URL,
        USER,
        PASSWORD,
        NONE,
    }

    private final Section section;
    private final String message;

    public LoginError(String message) {
        this(Section.NONE, message);
    }

    public LoginError(Section section, String message) {
        this.section = section;
        this.message = message;
    }

    public Section getSection() {
        return section;
    }

    public String getMessage() {
        return message;
    }

    public boolean isEmpty() {
        return message == null || message.isEmpty();
    }

    public static LoginError getError(Context context, int code, String defaultMessage) {
        switch (code) {
            case 401:
                return new LoginError(Section.USER, context.getString(R.string.error_incorrect_username_or_password));
            case 403:
            case 404:
                return new LoginError(Section.URL, context.getString(R.string.error_oc_not_found));
            case 405:
                return new LoginError(Section.URL, context.getString(R.string.ocnews_too_old));
            default:
                return new LoginError(defaultMessage);
        }
    }

    public static LoginError getError(Context context, Throwable t) {
        if (t instanceof UnknownHostException) {
            return new LoginError(Section.URL, context.getString(R.string.error_unknown_host));
        } else if (t instanceof SSLHandshakeException) {
            if (t.getCause() instanceof CertificateException) {
                return new LoginError(Section.URL, context.getString(R.string.untrusted_certificate));
            }
        }
        return new LoginError(t.getLocalizedMessage());
    }
}

File: app/src/main/java/email/schaal/ocreader/util/LoginError.kt
package email.schaal.ocreader.util

import android.content.Context
import com.squareup.moshi.Moshi
import email.schaal.ocreader.R
import retrofit2.HttpException
import java.io.IOException
import java.net.ConnectException
import java.net.UnknownHostException
import javax.net.ssl.SSLHandshakeException

/**
 * Turn login errors into human-readable strings
 */
class LoginError private constructor(val section: Section, val message: String, val throwable: Throwable? = null) {
    enum class Section {
        URL, USER, PASSWORD, NONE, UNKNOWN
    }

    private class ErrorMessage(val message: String)

    constructor(message: String) : this(Section.NONE, message)
    private constructor(throwable: Throwable) : this(Section.UNKNOWN, throwable.message ?: "", throwable)

    companion object {
        private fun getHttpError(context: Context, code: Int, e: HttpException): LoginError {
            return when (code) {
                401 -> LoginError(Section.USER, context.getString(R.string.error_access_forbidden))
                403, 404 -> LoginError(Section.URL, context.getString(R.string.error_oc_not_found))
                405 -> LoginError(Section.URL, context.getString(R.string.ncnews_too_old))
                else -> {
                    val message = e.response()?.errorBody()?.source()?.let { source ->
                        Moshi.Builder().build().adapter(ErrorMessage::class.java).fromJson(source)?.message }
                    LoginError("${context.getString(R.string.http_error, code)}${message.let { ": $it" }}")
                }
            }
        }

        fun getError(context: Context, t: Throwable): LoginError {
            return when (t) {
                is HttpException -> getHttpError(context, t.code(), t)
                is UnknownHostException -> LoginError(Section.URL, context.getString(R.string.error_unknown_host))
                is SSLHandshakeException -> LoginError(Section.URL, context.getString(R.string.untrusted_certificate))
                is ConnectException -> LoginError(Section.URL, context.getString(R.string.could_not_connect))
                is IOException -> return LoginError(Section.NONE, context.getString(R.string.ncnews_too_old))
                else -> LoginError(t)
            }
        }
    }

}